### ISSUE TITLE
Implement XLA-Rs for all math/model computation

### ISSUE THREAD

**Issue by BradenEverson** [BradenEverson/unda#8]
https://github.com/LaurentMazare/xla-rs

**Comment by Ebanflo42** [comment]
I was looking through your code and the xla-rs code, there are a handful of things which might need to be realigned. I thought maybe it would be helpful to at least create a roadmap:

1) `xla-rs` needs the proper version of `xla_extension` to be unzipped in its root directory, so either this has to be done manually on your local machine or there is a way to get cargo to do this automatically.

2) Using `Vec<Vec<...>>` to represent multidimensional arrays means dynamically sized arrays + non-contiguous storage = slow and RAM hungry. xla::Literal probably encapsulates the behavior that you are looking for here. It does not implement Send+Sync, but I'm not sure if that is a problem.

3) Binding optimizer state to individual layers is a nice idea, but I think the way you have done it severely limits the flexibility of the optimizer (what if I don't want to use Adam). I recommend checking out the `optax` library for some inspiration here.

4) One needs to think about how to actually build and execute the compute graphs. With your current network definition you only support sequential networks. The thing is skip connections are indispensable for NLP and CV, plus a more general way of constructing compute graphs is handy for designing recurrent networks. Based on the way `xla_builder` works it would probably be the easiest to structure to library in a TensorFlow 1 style, but this structure has been very unpopular since pytorch came out. I don't have a final solution to this but it's something important to think about.

I have some extra free time until next month, and I think one thing I will focus on is how to get automatic differentiation to work with `xla_rs`. I'm not sure how far I will get, if you want I can do this on a branch of your repo.

**Comment by BradenEverson** [comment]
> I was looking through your code and the xla-rs code, there are a handful of things which might need to be realigned. I thought maybe it would be helpful to at least create a roadmap:
> 
> 1. `xla-rs` needs the proper version of `xla_extension` to be unzipped in its root directory, so either this has to be done manually on your local machine or there is a way to get cargo to do this automatically.
> 2. Using `Vec<Vec<...>>` to represent multidimensional arrays means dynamically sized arrays + non-contiguous storage = slow and RAM hungry. xla::Literal probably encapsulates the behavior that you are looking for here. It does not implement Send+Sync, but I'm not sure if that is a problem.
> 3. Binding optimizer state to individual layers is a nice idea, but I think the way you have done it severely limits the flexibility of the optimizer (what if I don't want to use Adam). I recommend checking out the `optax` library for some inspiration here.
> 4. One needs to think about how to actually build and execute the compute graphs. With your current network definition you only support sequential networks. The thing is skip connections are indispensable for NLP and CV, plus a more general way of constructing compute graphs is handy for designing recurrent networks. Based on the way `xla_builder` works it would probably be the easiest to structure to library in a TensorFlow 1 style, but this structure has been very unpopular since pytorch came out. I don't have a final solution to this but it's something important to think about.
> 
> I have some extra free time until next month, and I think one thing I will focus on is how to get automatic differentiation to work with `xla_rs`. I'm not sure how far I will get, if you want I can do this on a branch of your repo.

On my local machine I was able to get the XLA_EXTENSION environment variable set up and it did seem to be working. I hadn't been able to do much yet just due to school aside from display GPU/CPU specs though.

Yeah I will be honest my whole matrix math side is very unoptimized/slow. My hope was to get something working and the look back into using an existing library, which is exactly what xla-rs would be doing for us hopefully!

For optimization too, I was planning on getting an enum type set up to choose optimization method(probably chosen in the during the fit method because I don't really see a need for different optimizers per layer), I'll check out optax though

I have seen a lot about compute graphs lately, would you recommend I rebuild some of the network side of the crate with this design pattern in mind? I don't really know too much about them aside from some basic stuff, is there a good resource I could find for implementation/general structure?

Thanks a lot for all of the tips man, if you'd like to mess around with xla_rs and getting automatic differentiation working, feel free to branch it off this repo. Let me know if you plan on getting started, and I'll start by looking more into graphs

**Comment by BradenEverson** [comment]
@Ebanflo42 I've made a new branch `compute-graph` where I'll play around with replacing the inner Layer/computations with tensors and operations on a computational graph. I've read a few research papers on it so far but am still a little new to the concept, so I may have a handful of questions that I may bombard you with at any moment

**Comment by Ebanflo42** [comment]
Awesome, I'm also doing a master's in math by the way, my extra free time comes from the fact that we have about 3 weeks until the next semester starts here :)

I was actually about to comment here right when you posted your second comment :sweat_smile:  I actually started a branch locally called `autodiff` where I plan on prototyping compute graphs, differentiating them, and converting them to XLA. I currently only know how to do the case where there is one output tensor, for more general cases I will have to think and read a bit more (in principle the loss is the only output you need for training, but its ideal to be able to look at hidden layer activations etc). I will push in a couple days and we can compare notes :)

For the optimizers, one case I can think of is transfer learning, where you have several pretrained layers followed by several untrained layers and you may, for example, want the learning rate on the pretrained layers to be much smaller. But this aspect might be something to consider further in the future, I will leave it to you to think about for now :)

**Comment by BradenEverson** [comment]
Sweet! I have no major background in math(I'm currently in a linear algebra class this semeseter so that's about where my experience ends) so it's great to have your skillset and knowledge available. That `autodiff` branch is a good idea, I'll let you primarily focus on the compute graphing aspect on this point in that case, but feel free to push regularly so I can poke around for fun :) I'll try a few basic things on my graph branch as well! 

That is true, in the case of transfer learning we could just set the learning_rate to be very low or even just 0. In the future you're definitely right that we may need to rework how optimizers are used per layer though.

Cheers! I hope your next few weeks are relatively stress free before your next semester :) and thanks for all the help thus far!
